File: macos/WaiComputer/WaiComputer/Features/Dictation/DictationLanguageStore.swift
import Foundation
import os

private let langLog = Logger(subsystem: "is.waiwai.computer.app", category: "dictation-language")

/// Source of truth for the user's dictation language preference.
///
/// Storage model is a `Set<String>` of BCP-47 codes persisted as a JSON array
/// in UserDefaults under `dictationLanguages`. Two modes derive from the set's
/// cardinality:
///
///   - **0 entries**: multilingual auto-detect. Wire tag is `""`; provider
///     adapters send the provider's explicit multilingual mode.
///   - **1 entry**: single-language mode — send that BCP-47 code as a hint
///     for the lowest possible latency on supported models.
///
/// Migration from the legacy single `transcriptionLanguage` string runs once
/// on first read. The legacy key stays in place so older builds that still
/// read it keep working until they update.
@MainActor
final class DictationLanguageStore: ObservableObject {
    static let userDefaultsKey = "dictationLanguages"
    static let legacyKey = "transcriptionLanguage"

    @Published private(set) var selectedLanguages: Set<String>

    init(defaults: UserDefaults = .standard) {
        let loaded = Self.loadOrMigrate(from: defaults)
        self.selectedLanguages = loaded
        Self.persist(loaded, defaults: defaults)
    }

    /// What we actually send to the upstream STT provider. Empty string for
    /// auto-detect; the single code for 1-language mode.
    var wireLanguageTag: String {
        if selectedLanguages.count == 1, let only = selectedLanguages.first {
            return only
        }
        return ""
    }

    /// True when the user has not picked any specific language — i.e. the
    /// model auto-detects across all supported languages.
    var isAutoDetect: Bool {
        selectedLanguages.isEmpty
    }

    func setLanguages(_ languages: Set<String>, defaults: UserDefaults = .standard) {
        let normalized = Self.normalizedSelection(languages)
        selectedLanguages = normalized
        persist(normalized, defaults: defaults)
        langLog.info("Dictation languages updated: \(normalized.sorted().joined(separator: ","), privacy: .public)")
    }

    func toggle(_ language: String, defaults: UserDefaults = .standard) {
        guard let normalized = Self.normalizedLanguage(language) else { return }
        if selectedLanguages == [normalized] {
            setAutoDetect(defaults: defaults)
        } else {
            setLanguages([normalized], defaults: defaults)
        }
    }

    func setAutoDetect(defaults: UserDefaults = .standard) {
        setLanguages([], defaults: defaults)
    }

    // MARK: - Persistence

    private func persist(_ languages: Set<String>, defaults: UserDefaults) {
        Self.persist(languages, defaults: defaults)
    }

    private static func persist(_ languages: Set<String>, defaults: UserDefaults) {
        let array = Array(languages).sorted()
        if let data = try? JSONEncoder().encode(array) {
            defaults.set(data, forKey: Self.userDefaultsKey)
        }
        // Mirror to legacy single-string key so older code paths reading
        // `transcriptionLanguage` keep working.
        let legacyValue = languages.count == 1 ? languages.first! : "multi"
        defaults.set(legacyValue, forKey: Self.legacyKey)
    }

    private static func loadOrMigrate(from defaults: UserDefaults) -> Set<String> {
        if let data = defaults.data(forKey: userDefaultsKey),
           let array = try? JSONDecoder().decode([String].self, from: data) {
            return normalizedSelection(Set(array))
        }
        // First-run migration: read legacy key.
        if let legacy = defaults.string(forKey: legacyKey) {
            switch legacy {
            case "multi", "":
                return []
            default:
                return normalizedSelection([legacy])
            }
        }
        return []  // Default to auto-detect if nothing stored.
    }

    private static func normalizedSelection(_ languages: Set<String>) -> Set<String> {
        let cleaned = Set(languages.compactMap(normalizedLanguage))
        guard cleaned.count == 1, let only = cleaned.first else { return [] }
        return [only]
    }

    private static func normalizedLanguage(_ language: String) -> String? {
        let normalized = language.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        return normalized.isEmpty || normalized == "multi" || normalized == "auto" ? nil : normalized
    }
}

@MainActor
enum DictationLanguageSelectionPolicy {
    static func providerLanguage(
        store: DictationLanguageStore?,
        defaults: UserDefaults = .standard
    ) -> String {
        if let store {
            return providerLanguage(fromWireTag: store.wireLanguageTag)
        }
        return providerLanguage(
            fromWireTag: defaults.string(forKey: DictationLanguageStore.legacyKey)
        )
    }

    static func providerLanguage(fromWireTag tag: String?) -> String {
        guard let normalized = normalizedProviderLanguage(tag) else {
            return "multi"
        }
        return normalized
    }

    private static func normalizedProviderLanguage(_ language: String?) -> String? {
        guard let language else { return nil }
        let normalized = language.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        return normalized.isEmpty || normalized == "multi" || normalized == "auto" ? nil : normalized
    }
}

enum DictationSessionConfigInvalidationPolicy {
    static func shouldClearVault(
        previousProvider: String?,
        previousModel: String?,
        nextProvider: String,
        nextModel: String
    ) -> Bool {
        guard let previousProvider, let previousModel else { return false }
        return previousProvider != nextProvider || previousModel != nextModel
    }
}

/// Static catalogue of languages exposed in the picker UI. Order is
/// frequency-of-use for the WaiComputer user base — English first, Russian second
/// (a top-2 user language), then a Whispr-Flow-style sweep of common picks.
struct DictationLanguageCatalog {
    struct Entry: Hashable, Identifiable {
        let code: String
        let englishName: String
        let nativeName: String
        var id: String { code }
    }

    static let all: [Entry] = [
        .init(code: "en", englishName: "English", nativeName: "English"),
        .init(code: "ru", englishName: "Russian", nativeName: "Русский"),
        .init(code: "es", englishName: "Spanish", nativeName: "Español"),
        .init(code: "de", englishName: "German", nativeName: "Deutsch"),
        .init(code: "fr", englishName: "French", nativeName: "Français"),
        .init(code: "it", englishName: "Italian", nativeName: "Italiano"),
        .init(code: "pt", englishName: "Portuguese", nativeName: "Português"),
        .init(code: "ja", englishName: "Japanese", nativeName: "日本語"),
        .init(code: "ko", englishName: "Korean", nativeName: "한국어"),
        .init(code: "hi", englishName: "Hindi", nativeName: "हिन्दी"),
        .init(code: "ar", englishName: "Arabic", nativeName: "العربية"),
        .init(code: "uk", englishName: "Ukrainian", nativeName: "Українська"),
        .init(code: "pl", englishName: "Polish", nativeName: "Polski"),
        .init(code: "nl", englishName: "Dutch", nativeName: "Nederlands"),
        .init(code: "tr", englishName: "Turkish", nativeName: "Türkçe"),
    ]

    static func entry(for code: String) -> Entry? {
        all.first { $0.code == code }
    }
}

File: desktop/WaiComputer.Core/Dictation/DictationLanguageStore.cs
using System.Linq;
using System.Text.Json;

namespace WaiComputer.Core.Dictation;

/// <summary>Key-value preferences port (Windows registry / Linux config file). Platforms implement.</summary>
public interface IPreferences
{
    string? Get(string key);
    void Set(string key, string value);
}

/// <summary>
/// Source of truth for the user's dictation language preference, porting the
/// macOS <c>DictationLanguageStore</c>. Stored as a JSON array of BCP-47 codes:
/// 0 entries = multilingual auto-detect (wire tag <c>""</c>); 1 entry =
/// single-language mode (that code). Migrates the legacy single
/// <c>transcriptionLanguage</c> string on first read and mirrors back to it.
/// </summary>
public sealed class DictationLanguageStore
{
    public const string PreferencesKey = "dictationLanguages";
    public const string LegacyKey = "transcriptionLanguage";

    private readonly IPreferences _prefs;
    private HashSet<string> _selected;

    public DictationLanguageStore(IPreferences prefs)
    {
        _prefs = prefs;
        _selected = LoadOrMigrate(prefs);
        Persist(_selected);
    }

    public IReadOnlySet<string> SelectedLanguages => _selected;

    /// <summary>Empty when the user lets the model auto-detect across all languages.</summary>
    public bool IsAutoDetect => _selected.Count == 0;

    /// <summary>What we send upstream: the single code, or <c>""</c> for auto-detect.</summary>
    public string WireLanguageTag => _selected.Count == 1 ? _selected.First() : string.Empty;

    public void SetLanguages(IEnumerable<string> languages)
    {
        _selected = NormalizedSelection(languages);
        Persist(_selected);
    }

    public void Toggle(string language)
    {
        var normalized = NormalizedLanguage(language);
        if (normalized is null)
        {
            return;
        }
        if (_selected.Count == 1 && _selected.Contains(normalized))
        {
            SetAutoDetect();
        }
        else
        {
            SetLanguages(new[] { normalized });
        }
    }

    public void SetAutoDetect() => SetLanguages(Array.Empty<string>());

    private void Persist(HashSet<string> languages)
    {
        var sorted = languages.OrderBy(x => x, StringComparer.Ordinal).ToArray();
        _prefs.Set(PreferencesKey, JsonSerializer.Serialize(sorted));
        // Mirror to the legacy single-string key so older code paths keep working.
        _prefs.Set(LegacyKey, languages.Count == 1 ? languages.First() : "multi");
    }

    private static HashSet<string> LoadOrMigrate(IPreferences prefs)
    {
        var raw = prefs.Get(PreferencesKey);
        if (raw is not null)
        {
            try
            {
                var array = JsonSerializer.Deserialize<string[]>(raw);
                if (array is not null)
                {
                    return NormalizedSelection(array);
                }
            }
            catch (JsonException) { /* fall through to legacy / default */ }
        }

        var legacy = prefs.Get(LegacyKey);
        if (legacy is not null)
        {
            return legacy is "multi" or "" ? new HashSet<string>() : NormalizedSelection(new[] { legacy });
        }

        return new HashSet<string>(); // default: auto-detect
    }

    private static HashSet<string> NormalizedSelection(IEnumerable<string> languages)
    {
        var cleaned = new HashSet<string>(languages.Select(NormalizedLanguage).OfType<string>(), StringComparer.Ordinal);
        return cleaned.Count == 1 ? cleaned : new HashSet<string>();
    }

    private static string? NormalizedLanguage(string language)
    {
        var normalized = language.Trim().ToLowerInvariant();
        return normalized is "" or "multi" or "auto" ? null : normalized;
    }
}

/// <summary>Maps the stored wire tag onto the provider's language parameter (auto-detect = "multi").</summary>
public static class DictationLanguageSelectionPolicy
{
    public static string ProviderLanguage(string? wireTag)
    {
        var normalized = wireTag?.Trim().ToLowerInvariant();
        return string.IsNullOrEmpty(normalized) || normalized is "multi" or "auto" ? "multi" : normalized;
    }
}

/// <summary>Whether a cached realtime session config must be discarded after a provider/model change.</summary>
public static class DictationSessionConfigInvalidationPolicy
{
    public static bool ShouldClearVault(string? previousProvider, string? previousModel, string nextProvider, string nextModel)
    {
        if (previousProvider is null || previousModel is null)
        {
            return false;
        }
        return previousProvider != nextProvider || previousModel != nextModel;
    }
}

/// <summary>A selectable dictation language (picker UI).</summary>
public sealed record DictationLanguage(string Code, string EnglishName, string NativeName);

/// <summary>Static catalogue of picker languages, ordered by WaiComputer usage frequency.</summary>
public static class DictationLanguageCatalog
{
    public static IReadOnlyList<DictationLanguage> All { get; } = new[]
    {
        new DictationLanguage("en", "English", "English"),
        new DictationLanguage("ru", "Russian", "Русский"),
        new DictationLanguage("es", "Spanish", "Español"),
        new DictationLanguage("de", "German", "Deutsch"),
        new DictationLanguage("fr", "French", "Français"),
        new DictationLanguage("it", "Italian", "Italiano"),
        new DictationLanguage("pt", "Portuguese", "Português"),
        new DictationLanguage("ja", "Japanese", "日本語"),
        new DictationLanguage("ko", "Korean", "한국어"),
        new DictationLanguage("hi", "Hindi", "हिन्दी"),
        new DictationLanguage("ar", "Arabic", "العربية"),
        new DictationLanguage("uk", "Ukrainian", "Українська"),
        new DictationLanguage("pl", "Polish", "Polski"),
        new DictationLanguage("nl", "Dutch", "Nederlands"),
        new DictationLanguage("tr", "Turkish", "Türkçe"),
    };

    public static DictationLanguage? Entry(string code) => All.FirstOrDefault(e => e.Code == code);
}
